### PR TITLE
Hardware serial

### PR DESCRIPTION
**Pull request by RuthgerD**
Towards arduino compliance :)
WIP

### CODE REVIEW

**Comment by RuthgerD** [comment]
Drafted for easier review

File: include/ardpolyfills/HardwareSerial.h [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+    void begin(unsigned long baud) { begin(baud, SERIAL_8N1); }
+    void begin(unsigned long, std::uint8_t);
+    void end();
+
+    int available() override;
+    int read() override;
+    int peek() override;
+
+    int availableForWrite() override;
```
Comment (AeroStun) [inline include/ardpolyfills/HardwareSerial.h:64]: Should be `[[nodiscard]]`

File: include/ardpolyfills/HardwareSerial.h [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+    BoardData board_data;
+
+    void begin(unsigned long baud) { begin(baud, SERIAL_8N1); }
+    void begin(unsigned long, std::uint8_t);
+    void end();
+
+    int available() override;
+    int read() override;
+    int peek() override;
```
Comment (AeroStun) [inline include/ardpolyfills/HardwareSerial.h:62]: Should be `[[nodiscard]]`

File: include/ardpolyfills/Print.h [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        const auto beg = buffer;
+        while (size-- && write(*buffer++))
+            ;
+        return std::distance(beg, buffer);
+    }
+    inline std::size_t write(const char* str) {
+        if (!str)
+            return 0;
+        return write(str, std::strlen(str));
+    }
```
Comment (AeroStun) [inline include/ardpolyfills/Print.h:50]: De-inline and define out-of-line in a separate TU

File: include/ardpolyfills/Stream.h [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    virtual int available() = 0;
+    virtual int read() = 0;
+    virtual int peek() = 0;
+
+    Stream() = default;
+
+    bool find(char target) { return find(&target, 0); }
+    bool find(char* target, size_t length) { return findUntil(target, length, NO_IGNORE_CHAR); }
+    bool findUntil(char target, char terminal) { return findUntil(&target, 0, NO_IGNORE_CHAR); }
+    bool findUntil(char* target, int length, char terminal);
```
Comment (AeroStun) [inline include/ardpolyfills/Stream.h:45]: Should be `[[nodiscard]]`, `const` and `noexcept`

File: modules/ardrivo/include/ardrivo/BoardData.hxx [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+ */
+
+#ifndef SMARTCAR_EMUL_BOARDDATA_HXX
+#define SMARTCAR_EMUL_BOARDDATA_HXX
+
+#include <vector>
+#include <deque>
+
+extern bool (*write_byte)(unsigned char);
+extern size_t (*write_buf)(const unsigned char*, size_t);
```
Comment (AeroStun) [inline modules/ardrivo/include/ardrivo/BoardData.hxx:26]: Move these in `BoardData`

**Comment by AeroStun** [comment]
Also missing mutex locking and `BoardData`'s instance be accessible through a global pointer

**Comment by RuthgerD** [comment]
PR technically depends on #18 so this branch needs rebase once thats complete

File: modules/ardrivo/src/Stream.cxx [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+    return value;
+}
+
+size_t Stream::readBytesUntil(char terminator, char* buffer, int length) {
+    size_t index = 0;
+    while (index < length) {
+        int c = read();
+        if (c < 0 || c == terminator)
+            break;
+        *buffer++ = (char)c;
```
Comment (AeroStun) [inline modules/ardrivo/src/Stream.cxx:112]: Narrowing cast; please use `static_cast`

File: modules/ardrivo/src/Stream.cxx [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+        read();
+        c = peek();
+    } while (std::isdigit(c) || (c == '.' && !isFraction) || c == ignore);
+
+    if (isNegative)
+        value = -value;
+    if (isFraction)
+        return value * fraction;
+    else
+        return value;
```
Comment (AeroStun) [inline modules/ardrivo/src/Stream.cxx:79]: Use ternary expression or first multiply if necessary and then return unconditionally

**Pull request closed without merging**